### ISSUE TITLE
bGMM hangs indefinitely for one site, possibly due to zero M-values

### ISSUE THREAD

**Issue by ning-y** [BenjaminPlanterose/UMtools#2]
UMtools::bGMM hangs indefinitely for one site in my dataset, cg22855933. This site has many zero M values:

<div align='center'>
<img src='https://user-images.githubusercontent.com/22320312/200107172-20bf50b1-0f80-4aac-9ab6-2faeb8ec8411.png'/>
</div>

A SIGNINT (control-c) has not effect, as in the R session continues to hang. `R.utils::withTimeout` does not work either, as in the command continues to hang indefinitely even pass the specified timeout duration. I am running interactive R on the command line, so by 'hangs indefinitely' I mean that the prompt "> " does not reappear, and I cannot enter any more commands into the R session.

Due to lack of time I am not able to pinpoint which statement in UMtools::bGMM causes this behavior, though I have saved the UMtools::GR_to_UM M and U values for this site as RDS, and it is attached: [cg22855933.tar.gz](https://github.com/BenjaminPlanterose/UMtools/files/9943258/cg22855933.tar.gz)

p.s. I understand that this is a low-quality site, but it would be good to resolve this in the context of batch processing many sites. For now I have resolved this by filtering out any sites with zero M or U values.

**Comment by BenjaminPlanterose** [comment]
Dear ning-y, 
thank you very much for reporting this issue. This can be quickly fixed by adding some almost zero random noise as follows:

```r
library(UMtools)
beta = readRDS('cg22855933.rds')
res = bGMM(M = beta$M + rnorm(n = length(beta$M), mean = 0, sd = 0.1), U = beta$U + rnorm(n = length(beta$U), mean = 0, sd = 0.1), CpG = 'cg22855933', K = 2)
table(res)
# blue  red 
# 277   87 
```
The reason the Gaussian mixture model fitting fails is because it estimates the covariance matrix for each cluster as part of an EM algorithm. But when all the values are equal to zero in one cluster, the variance is also equal to zero, generating a singular covariance matrix for that cluster. Somehow, the implementation by `EMCluster::assign.class` does not tolerate this and ends up in an unfortunate infinite loop (i.e. imagine gradient descent with a learning rate of zero). Since this implementation runs in C++, R cannot kill this process swiftly, resulting in a frozen session. 

In my experience with DNA methylation microarray data, I have never seen so many 0-valued fluorescence intensities. Are these processed in any way? Or are you using some low-quality/quantity template?

In any case, I will incorporate this fix into the repository in the coming weeks. In the meanwhile, I recommend that you define the following function to overrun that from UMtools:
```r
bGMM <- function(M, U, CpG, K, stable.solution = TRUE, min.n = NULL, 
                 min.n.iter = 2000, method = "em.EM", EMC = .EMC, transform = TRUE)
{
  m = M[CpG, ]; m = m + rnorm(n = length(m), mean = 0, sd = 0.1)
  u = U[CpG, ]; u = u + rnorm(n = length(u), mean = 0, sd = 0.1)
  
  if (transform) {
    df = data.frame(x = m/(u + m + 100), y = log2(u + m + 100))
    df$y = (df$y - min(df$y))/(max(df$y))
  }
  else {
    df = data.frame(x = m, y = u)
  }
  ret <- EMCluster::init.EM(x = df, nclass = K, method = "em.EM", 
                            min.n.iter = min.n.iter, lab = NULL, EMC = .EMC, stable.solution = stable.solution, 
                            min.n = min.n)
  class_i <- EMCluster::assign.class(df, ret, return.all = FALSE)$class
  if (K == 1 | K > 4) 
  {
    if (K > 8) {
      stop("Too many K and not enough colours")
    }
    pal = brewer.pal(K, "Dark2")
    df = data.frame(x = m, y = u)
    col = as.factor(class_i)
    levels(col) = c("Dark green", "Orange", "Purple", "Pink", 
                    "Light green", "Yellow", "Brown", "Gray")[1:K]
    plot(df$x, df$y, col = as.character(col), pch = 19, main = CpG, 
         xlab = "M", ylab = "U", xlim = c(0, max(df$x) + 100), 
         ylim = c(0, max(df$y) + 100))
    return(as.character(col))
  }
  else if (K == 2) {
    df = data.frame(x = m, y = u)
    coefA = aggregate(. ~ class_i, df[c("x")], mean)$x
    coefB = aggregate(. ~ class_i, df[c("y")], mean)$y
    pos1 = which.min((coefA + coefB)/2)
    pos2 = which.max((coefA + coefB)/2)
    real_order <- c(pos1, pos2)
    class_i <- factor(class_i, levels = c(1, 2, 3, 4))
    levels(class_i) <- as.character(match(1:4, real_order))
    col_vec <- as.vector(class_i)
    out = col_vec
    col_vec[col_vec == "1"] <- scales::alpha("dodgerblue3", 
                                                          0.5)
    col_vec[col_vec == "2"] <- scales::alpha("brown3", 0.5)
    out[out == "1"] <- "blue"
      out[out == "2"] <- "red"
        plot(df$x, df$y, col = col_vec, pch = 19, main = CpG, 
             xlab = "M", ylab = "U", xlim = c(0, max(df$x) + 100), 
             ylim = c(0, max(df$y) + 100))
        return(out)
  }
  else if (K == 3) {
    df = data.frame(x = m, y = u)
    coefs <- numeric()
    lm.mod <- lm(df$y[class_i == 1] ~ df$x[class_i == 1])
    coefs[1] <- unname(lm.mod$coefficients[2])
    lm.mod <- lm(df$y[class_i == 2] ~ df$x[class_i == 2])
    coefs[2] <- unname(lm.mod$coefficients[2])
    lm.mod <- lm(df$y[class_i == 3] ~ df$x[class_i == 3])
    coefs[3] <- unname(lm.mod$coefficients[2])
    class_i <- factor(class_i, levels = c(1, 2, 3))
    levels(class_i) <- as.character(match(1:3, order(coefs, 
                                                     decreasing = T)))
    col_vec <- as.vector(class_i)
    out = col_vec
    col_vec[col_vec == "1"] <- scales::alpha("dodgerblue3", 0.5)
    col_vec[col_vec == "2"] <- scales::alpha("darkmagenta", 0.5)
    col_vec[col_vec == "3"] <- scales::alpha("brown3", 0.5)
    out[out == "1"] <- "blue"
      out[out == "2"] <- "purple"
        out[out == "3"] <- "red"
          plot(df$x, df$y, col = col_vec, pch = 19, main = CpG, 
               xlab = "M", ylab = "U", xlim = c(0, max(df$x) + 100), 
               ylim = c(0, max(df$y) + 100))
          return(out)
  }
  else if (K == 4) {
    df = data.frame(x = m, y = u)
    df_norm = df/data.frame(x = rep(max(df$x), nrow(df)), 
                            y = rep(max(df$y), nrow(df)))
    df_norm$label = class_i
    coefA = aggregate(. ~ label, df_norm[c("x", "label")], 
                      mean)$x
    coefB = aggregate(. ~ label, df_norm[c("y", "label")], 
                      mean)$y
    pos1 = which.min((coefA + coefB)/2)
    coefA[pos1] = NA
    coefB[pos1] = NA
    pos4 = which.min(abs(coefA - coefB))
    coefA[pos4] = NA
    coefB[pos4] = NA
    pos2 = which.max(coefB)
    pos3 = which.max(coefA)
    real_order <- c(pos1, pos2, pos3, pos4)
    class_i <- factor(class_i, levels = c(1, 2, 3, 4))
    levels(class_i) <- as.character(match(1:4, real_order))
    col_vec <- as.vector(class_i)
    out = col_vec
    col_vec[col_vec == "1"] <- scales::alpha("black", 0.5)
    col_vec[col_vec == "2"] <- scales::alpha("dodgerblue3", 
                                                          0.5)
    col_vec[col_vec == "3"] <- scales::alpha("brown3", 0.5)
    col_vec[col_vec == "4"] <- scales::alpha("darkmagenta", 
                                                          0.5)
    out[out == "1"] <- "blue"
      out[out == "2"] <- "purple"
        out[out == "3"] <- "red"
          out[out == "4"] <- "red"
            plot(df$x, df$y, col = col_vec, pch = 19, main = CpG, 
                 xlab = "M", ylab = "U", xlim = c(0, max(df$x) + 100), 
                 ylim = c(0, max(df$y) + 100))
            return(out)
  }
}
```



Thanks again.

Best,
Ben.
